frontend/src/js/controllers/challengeCtrl.js

```diff
@@ -2878,6 +2878,86 @@
             }
         };
 
+        vm.editchallengeTagDialog = function(ev) {
+            vm.tags = [];
+            for (var i = 0; i < vm.page.list_tags.length; i++) {
+                vm.tags[i] = vm.page.list_tags[i];
+            }
+            vm.domainoptions = vm.domain_choices();
+            $mdDialog.show({
+                scope: $scope,
+                preserveScope: true,
+                targetEvent: ev,
+                templateUrl: 'dist/views/web/challenge/edit-challenge/edit-challenge-tag.html',
+                escapeToClose: false
+            });
+        };
+
+        vm.editChallengeTag = function(editChallengeTagDomainForm) {
+            var new_tags;
+            if (!editChallengeTagDomainForm) {
+                $mdDialog.hide();
+                return;
+            }
+            new_tags = typeof vm.tags === 'string'
+                ? vm.tags.split(",").map(item => item.trim())
+                : vm.page.list_tags.map(tag => tag);
+
+            if (typeof vm.tags === 'string' && (new_tags.length > 4 || new_tags.some(tag => tag.length > 15))) {
+                $rootScope.notify("error", "Invalid tags! Maximum 4 tags are allowed, and each tag must be 15 characters or less.");
+                return;
+            }
+            parameters.url = "challenges/challenge/" + vm.challengeId + "/update_challenge_tags_and_domain/";
+            parameters.method = 'PATCH';
+            parameters.data = {
+                "list_tags": new_tags,
+                "domain": vm.domain
+            };
+            parameters.callback = {
+                onSuccess: function(response) {
+                    var status = response.status;
+                    utilities.hideLoader();
+                    if (status === 200) {
+                        $rootScope.notify("success", "The challenge tags and domain is successfully updated!");
+                        $state.reload();
+                        $mdDialog.hide();
+                    }
+                },
+                onError: function(response) {
+                    utilities.hideLoader();
+                    $mdDialog.hide();
+                    var error = response.data;
+                    $rootScope.notify("error", error);
+                }
+            };
+            utilities.showLoader();
+            utilities.sendRequest(parameters);
+        };
+
+        vm.domain_choices = function() {
+            var domain_choices = [];
+            parameters.url = "challenges/challenge/get_domain_choices/";
+            parameters.method = 'GET';
+            parameters.data = {};
+            parameters.callback = {
+                onSuccess: function(response) {
+                    var details = response.data;
+                    for (var i = 0; i < details.length; i++) {
```

Comment:
optimise this?
do we really need "details" variable? domain_choices can be assigned response.data?

Reply:
We can use domain_choices, i will change it